prism/src/main/java/org/apache/falcon/resource/AbstractEntityManager.java

```diff
@@ -354,6 +357,12 @@ protected APIResult update(Entity newEntity, String type, String entityName, Boo
             }
 
             configStore.update(entityType, newEntity);
+            // for Data Sources, check always if dependant feeds are already upgraded and upgrade accordingly
+            if (entityType.equals(EntityType.DATASOURCE)) {
+                ConfigurationStore.get().cleanupUpdateInit();
+                releaseEntityLocks(entityName, tokenList);
+                updateDatasourceDependents(entityName, skipDryRun);
```

Comment:
If it is a single-step update, the dependents are updated before the configStore is updated. Else, if the updateDatasourceDependents fails midway, there is no way to recover.